pkg/addons/custom/custom.go

```diff
@@ -57,43 +54,72 @@ func (c *Custom) HostPreflights() (*v1beta2.HostPreflightSpec, error) {
 	return nil, nil
 }
 
-func (c *Custom) Apply(ctx context.Context) error {
+// GenerateHelmConfig generates the helm config for all the embedded charts.
+// and writes the charts to the disk.
+func (c *Custom) GenerateHelmConfig(ctx *cli.Context, isUpgrade bool) ([]dig.Mapping, error) {
+
+	chartConfigs := []dig.Mapping{}
+
 	exe, err := os.Executable()
 	if err != nil {
-		return fmt.Errorf("unable to get executable path: %w", err)
+		return chartConfigs, fmt.Errorf("unable to get executable path: %w", err)
 	}
 	opts, err := hembed.ReadEmbedOptionsFromBinary(exe)
 	if err != nil {
-		return fmt.Errorf("unable to read embed options: %w", err)
+		return chartConfigs, fmt.Errorf("unable to read embed options: %w", err)
 	} else if opts == nil {
-		c.logger("No embed charts found, skipping custom addons.")
-		return nil
+		logrus.Warn("No embed charts found, skipping custom addons.")
+		return chartConfigs, nil
 	}
+
 	for _, chart := range opts.Charts {
-		if err := c.applyOne(ctx, chart); err != nil {
-			return fmt.Errorf("unable to apply chart: %w", err)
+
+		chartData, err := loader.LoadArchive(chart.ChartReader())
+		if err != nil {
+			return chartConfigs, fmt.Errorf("unable to load chart archive: %w", err)
 		}
-	}
-	return nil
-}
 
-func (c *Custom) applyOne(ctx context.Context, ochart hembed.HelmChart) error {
-	chart, err := loader.LoadArchive(ochart.ChartReader())
-	if err != nil {
-		return fmt.Errorf("unable to load chart archive: %w", err)
-	}
-	if c.chartHasBeenDisabled(chart) {
-		c.logger("Skipping disabled addon %s", chart.Name())
-		return nil
-	}
-	var values map[string]interface{}
-	if len(ochart.Values) > 0 {
-		values = make(map[string]interface{})
-		if err := yaml.Unmarshal([]byte(ochart.Values), &values); err != nil {
-			return fmt.Errorf("unable to unmarshal values: %w", err)
+		if c.chartHasBeenDisabled(chartData) {
+			logrus.Infof("skipping disabled chart %s", chartData.Name())
+			continue
 		}
+
+		chartName := strings.ToLower(chartData.Name())
+		chartFile := fmt.Sprintf("%s-%s.tgz", chartName, chartData.Metadata.Version)
+		dstpath := filepath.Join(defaults.HelmChartSubDir(), chartFile)
+
+		chartConfig := dig.Mapping{
+			"name":      chartName,
+			"namespace": c.namespace,
+			"version":   chartData.Metadata.Version,
+		}
+
+		chartConfig["chartName"] = dstpath
+
+		chartConfig["values"] = chart.Values
+
+		dst, err := os.Create(dstpath)
+
+		defer func() {
+			if err := dst.Close(); err != nil {
+				logrus.Errorf("error closing file %s: %s", dstpath, err)
+			}
+		}()
+
+		if err != nil {
```

Comment:
this is a bit off, if err is not nil then dst is and we are calling close on it in defer.